File: scala-3.6/tax-system/src/test/scala/com/rbleggi/taxsystem/service/TaxCalculatorTest.scala
package com.rbleggi.taxsystem.service

import com.rbleggi.taxsystem.model.{Book, Electronic, State, TaxRule}
import org.scalatest.funsuite.AnyFunSuite

class TaxCalculatorTest extends AnyFunSuite {

  val laptop = Electronic("Laptop", 1500.0)
  val book = Book("Book", 30.0)

  val california = State("California", "CA")
  val texas = State("Texas", "TX")

  val taxRules = List(
    TaxRule(california, laptop, 2025, 8.5),
    TaxRule(texas, laptop, 2025, 6.25),
    TaxRule(california, book, 2025, 0.0)
  )

  val taxCalculator = new TaxCalculator(taxRules)

  test("Calculate total price for Laptop in California") {
    assert(taxCalculator.calculateTotalPrice(laptop, california, 2025) === 1627.5)
  }

  test("Calculate total price for Laptop in Texas") {
    assert(taxCalculator.calculateTotalPrice(laptop, texas, 2025) === 1593.75)
  }

  test("Calculate total price for Book in California (tax-free)") {
    assert(taxCalculator.calculateTotalPrice(book, california, 2025) === 30.0)
  }

  test("Return product price when no tax rule exists") {
    val unknownState = State("New York", "NY")
    assert(taxCalculator.calculateTotalPrice(laptop, unknownState, 2025) === 1500.0)
  }
}

File: kotlin/tax-system/src/test/kotlin/com/rbleggi/taxsystem/service/TaxCalculatorTest.kt
package com.rbleggi.taxsystem.service

import com.rbleggi.taxsystem.model.Book
import com.rbleggi.taxsystem.model.Electronic
import com.rbleggi.taxsystem.model.State
import com.rbleggi.taxsystem.model.TaxRule
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class TaxCalculatorTest {

    private val laptop = Electronic(name = "Laptop", price = 1500.0)
    private val book = Book(name = "Book", price = 30.0)

    private val california = State(name = "California", code = "CA")
    private val texas = State(name = "Texas", code = "TX")

    private val taxRules = listOf(
        TaxRule(state = california, product = laptop, year = 2025, taxRate = 8.5),
        TaxRule(state = texas, product = laptop, year = 2025, taxRate = 6.25),
        TaxRule(state = california, product = book, year = 2025, taxRate = 0.0)
    )

    private val taxCalculator = TaxCalculator(taxRules)

    @Test
    fun `should calculate total price for Laptop in California`() {
        val totalPrice = taxCalculator.calculateTotalPrice(laptop, california, 2025)
        assertEquals(1627.5, totalPrice, 0.01)
    }

    @Test
    fun `should calculate total price for Laptop in Texas`() {
        val totalPrice = taxCalculator.calculateTotalPrice(laptop, texas, 2025)
        assertEquals(1593.75, totalPrice, 0.01)
    }

    @Test
    fun `should calculate total price for Book in California`() {
        val totalPrice = taxCalculator.calculateTotalPrice(book, california, 2025)
        assertEquals(30.0, totalPrice, 0.01)
    }

    @Test
    fun `should return product price when no tax rule exists`() {
        val unknownState = State(name = "New York", code = "NY")
        val totalPrice = taxCalculator.calculateTotalPrice(laptop, unknownState, 2025)
        assertEquals(1500.0, totalPrice, 0.01)
    }
}
